File: sdk/php/src/Internal/Infra/TopicManager.php
<?php

namespace KuCoin\UniversalSDK\Internal\Infra;

class TopicManager
{
    /** @var array<string, CallbackManager> $topicPrefix */
    private $topicPrefix = [];

    public function getCallbackManager($topic): CallbackManager
    {
        $parts = explode(':', $topic);
        $prefix = (count($parts) === 2 && $parts[1] !== 'all') ? $parts[0] : $topic;

        if (!isset($this->topicPrefix[$prefix])) {
            $this->topicPrefix[$prefix] = new CallbackManager($topic);
        }

        return $this->topicPrefix[$prefix];
    }

    /**
     * @param callable $func function(string $key, CallbackManager $value): bool
     */
    public function range(callable $func)
    {
        foreach ($this->topicPrefix as $key => $value) {
            if (!$func($key, $value)) {
                break;
            }
        }
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/TopicManager.java
package com.kucoin.universal.sdk.internal.infra;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.BiConsumer;

final class TopicManager {

  private final ConcurrentMap<String, CallbackManager> map = new ConcurrentHashMap<>();

  CallbackManager getCallbackManager(String topic) {
    String[] parts = topic.split(":", 2);
    String prefix = topic;
    if (parts.length == 2 && !"all".equals(parts[1])) {
      prefix = parts[0];
    }

    return map.computeIfAbsent(prefix, CallbackManager::new);
  }

  void forEach(BiConsumer<String, CallbackManager> action) {
    map.forEach(action);
  }

  boolean isEmpty() {
    return map.isEmpty();
  }
}
